src/storage/src/hummock/store/version.rs

```diff
@@ -213,44 +211,25 @@ impl HummockReadVersion {
                 // the batch id of imm at the front
                 StagingData::ImmMem(imm) => self.staging.imm.push_front(imm),
                 StagingData::Sst(staging_sst) => {
-                    // TODO: enable this stricter check after each streaming table owns a read
-                    // version. assert!(self.staging.imm.len() >=
-                    // staging_sst.imm_ids.len()); assert!(staging_sst
-                    //     .imm_ids
-                    //     .is_sorted_by(|batch_id1, batch_id2| batch_id2.partial_cmp(batch_id1)));
-                    // assert!(
-                    //     check_subset_preserve_order(
-                    //         staging_sst.imm_ids.iter().cloned(),
-                    //         self.staging.imm.iter().map(|imm| imm.batch_id()),
-                    //     ),
-                    //     "the imm id of staging sstable info not preserve the imm order. staging
-                    // sst imm ids: {:?}, current imm ids: {:?}",
-                    //     staging_sst.imm_ids.iter().collect_vec(),
-                    //     self.staging.imm.iter().map(|imm| imm.batch_id()).collect_vec()
-                    // );
-                    // for clear_imm_id in staging_sst.imm_ids.iter().rev() {
-                    //     let item = self.staging.imm.back().unwrap();
-                    //     assert_eq!(*clear_imm_id, item.batch_id());
-                    //     self.staging.imm.pop_back();
-                    // }
-
-                    debug_assert!(
-                        check_subset_preserve_order(
-                            staging_sst.imm_ids.iter().cloned().sorted(),
-                            self.staging.imm.iter().map(|imm| imm.batch_id()).sorted()
-                        ),
-                        "the set of imm ids in the staging_sst {:?} is not a subset of current staging imms {:?}",
-                        staging_sst.imm_ids.iter().cloned().sorted().collect_vec(),
-                        self.staging.imm.iter().map(|imm| imm.batch_id()).sorted().collect_vec(),
-                    );
-
-                    let imm_id_set: HashSet<ImmId> =
-                        HashSet::from_iter(staging_sst.imm_ids.iter().cloned());
-                    self.staging
-                        .imm
-                        .retain(|imm| !imm_id_set.contains(&imm.batch_id()));
+                    let staging_imm_ids_from_sst: HashSet<u64> =
+                        staging_sst.imm_ids.iter().cloned().collect();
+                    let staging_imm_ids_from_imms: HashSet<u64> =
+                        self.staging.imm.iter().map(|imm| imm.batch_id()).collect();
+                    let intersection =
+                        staging_imm_ids_from_imms.intersection(&staging_imm_ids_from_sst);
+
+                    // check intersection order of staging_imm
+                    let staging_imm_ids_from_sst_vec =
```

Comment:
The commented code is not correct. I realized that it was incorrect when I was implementing the uploader and therefore I commented the code and save for later reference. But the check logic itself is not correct.

First, the `staging_sst.imm_ids` is not ordered, because generating a new imm and adding imm to the channel is not atomic, and therefore it may happen that an imm with larger (later generated) batch id is added to the channel earlier than the one with smaller batch id. However, the staging.imm ids in a single read version is always sorted.

Second, the property we want to preserve is even stricter. It should be, the intersection between the two imm id set are always a continuous part at the back of `self.staging.imm`, and the relative order of this intersection is preserved in both lists. The current implementation has ensured that the intersection between the two imm id set are always a continuous part at the back of `self.staging.imm`. It leverages the property that the staging.imm ids in a single read version is always sorted. Now we sort the intersection part, and while iterating over the sorted intersection part, we keep `pop_back` and assert that popped imm id is the same as the imm id in this iteration. In this way, we can ensure that the intersection is always a continuous part at the back. However, it seems that we didn't check that the order is preserved yet.